File: experiments/android-readplace-poc/core/src/main/kotlin/com/readplace/poc/core/TokenStore.kt
package com.readplace.poc.core

/**
 * A minimal key/value seam so the core stays platform-free. The app backs this
 * with `SharedPreferences`; tests back it with an in-memory map. (A production app
 * would keep tokens in the Android Keystore / EncryptedSharedPreferences.)
 */
interface KeyValueStore {
	fun getString(key: String): String?
	fun putString(key: String, value: String)
	fun remove(key: String)
}

/** OAuth tokens issued by the server's `/oauth/token` endpoint. */
data class OAuthTokens(val accessToken: String, val refreshToken: String)

/**
 * Persists tokens and the active base URL. On Android the launcher app (which signs
 * in) and the share-target Activity (which saves) run in the same app sandbox, so a
 * single backing store is shared between them — no iOS-style App Group is needed.
 */
class TokenStore(private val store: KeyValueStore) {
	private object Key {
		const val ACCESS_TOKEN = "oauth.accessToken"
		const val REFRESH_TOKEN = "oauth.refreshToken"
		const val BASE_URL = "config.baseURL"
	}

	val tokens: OAuthTokens?
		get() {
			val access = store.getString(Key.ACCESS_TOKEN) ?: return null
			val refresh = store.getString(Key.REFRESH_TOKEN) ?: return null
			return OAuthTokens(access, refresh)
		}

	fun save(tokens: OAuthTokens) {
		store.putString(Key.ACCESS_TOKEN, tokens.accessToken)
		store.putString(Key.REFRESH_TOKEN, tokens.refreshToken)
	}

	fun updateAccessToken(accessToken: String, refreshToken: String?) {
		store.putString(Key.ACCESS_TOKEN, accessToken)
		refreshToken?.let { store.putString(Key.REFRESH_TOKEN, it) }
	}

	fun clear() {
		store.remove(Key.ACCESS_TOKEN)
		store.remove(Key.REFRESH_TOKEN)
	}

	var baseUrl: String
		get() = store.getString(Key.BASE_URL) ?: AppConfig.DEFAULT_BASE_URL
		set(value) = store.putString(Key.BASE_URL, value)

	val isLoggedIn: Boolean get() = tokens != null
}

File: projects/ios-readplace-poc/Shared/TokenStore.swift
import Foundation

/// OAuth tokens issued by the server's `/oauth/token` endpoint.
struct OAuthTokens: Equatable {
	let accessToken: String
	let refreshToken: String
}

/// Persists tokens and the active base URL in the shared App Group so the
/// app and the share extension agree on identity and server.
///
/// A POC-grade store: UserDefaults in the shared container. (A production app
/// would keep tokens in the Keychain with a shared access group.)
struct TokenStore {
	private let defaults: UserDefaults

	private enum Key {
		static let accessToken = "oauth.accessToken"
		static let refreshToken = "oauth.refreshToken"
		static let baseURL = "config.baseURL"
	}

	init() {
		let group = TokenStore.resolvedAppGroupId
		/// Falls back to standard defaults so the app still runs where the App
		/// Group entitlement is missing — the share extension simply won't see
		/// the token until the group is enabled.
		defaults = UserDefaults(suiteName: group) ?? .standard
		NSLog("[ReadplacePOC] TokenStore group=\(group) shared=\(UserDefaults(suiteName: group) != nil)")
	}

	/// Injectable backing store for tests.
	init(defaults: UserDefaults) {
		self.defaults = defaults
	}

	/// The App Group id this process is actually entitled to. A sideloader
	/// (AltStore/Sideloadly) may rewrite the declared id when re-signing, so we
	/// read it back from the embedded provisioning profile rather than trusting
	/// the compile-time constant. Falls back to the constant when unavailable.
	static let resolvedAppGroupId: String = {
		guard
			let url = Bundle.main.url(forResource: "embedded", withExtension: "mobileprovision"),
			let data = try? Data(contentsOf: url),
			let raw = String(data: data, encoding: .isoLatin1),
			let start = raw.range(of: "<?xml"),
			let end = raw.range(of: "</plist>")
		else { return AppConfig.appGroupId }
		let plistString = String(raw[start.lowerBound..<end.upperBound])
		guard
			let plistData = plistString.data(using: .isoLatin1),
			let plist = try? PropertyListSerialization.propertyList(from: plistData, format: nil) as? [String: Any],
			let entitlements = plist["Entitlements"] as? [String: Any],
			let groups = entitlements["com.apple.security.application-groups"] as? [String],
			let group = groups.first
		else { return AppConfig.appGroupId }
		return group
	}()

	var tokens: OAuthTokens? {
		guard
			let access = defaults.string(forKey: Key.accessToken),
			let refresh = defaults.string(forKey: Key.refreshToken)
		else { return nil }
		return OAuthTokens(accessToken: access, refreshToken: refresh)
	}

	func save(_ tokens: OAuthTokens) {
		defaults.set(tokens.accessToken, forKey: Key.accessToken)
		defaults.set(tokens.refreshToken, forKey: Key.refreshToken)
	}

	func updateAccessToken(_ accessToken: String, refreshToken: String?) {
		defaults.set(accessToken, forKey: Key.accessToken)
		if let refreshToken { defaults.set(refreshToken, forKey: Key.refreshToken) }
	}

	func clear() {
		defaults.removeObject(forKey: Key.accessToken)
		defaults.removeObject(forKey: Key.refreshToken)
	}

	var baseURL: String {
		get { defaults.string(forKey: Key.baseURL) ?? AppConfig.defaultBaseURL }
		nonmutating set { defaults.set(newValue, forKey: Key.baseURL) }
	}

	var isLoggedIn: Bool { tokens != nil }
}
